tss-esapi/src/structures/buffers/public/rsa.rs

```diff
@@ -0,0 +1,361 @@
+            }
+
+            if self.is_decryption_key && rsa_scheme.algorithm() != RsaSchemeAlgorithm::Null {
+                error!("Invalid rsa scheme algorithm provided with 'restricted' and 'is_decryption_key' set to true");
+                return Err(Error::local_error(WrapperErrorKind::InconsistentParams));
+            }
+        } else {
+            if self.is_decryption_key
+                && self.is_signing_key
+                && rsa_scheme.algorithm() != RsaSchemeAlgorithm::Null
+            {
+                error!("Invalid rsa scheme algorithm provided with 'restricted' set to false and 'is_decryption_key' and 'is_signing_key' set to true");
+                return Err(Error::local_error(WrapperErrorKind::InconsistentParams));
+            }
+            if self.is_signing_key
+                && rsa_scheme.algorithm() != RsaSchemeAlgorithm::RsaPss
+                && rsa_scheme.algorithm() != RsaSchemeAlgorithm::RsaSsa
+                && rsa_scheme.algorithm() != RsaSchemeAlgorithm::Null
+            {
+                error!("Invalid rsa scheme algorithm provided with 'restricted' set to false and 'is_signing_key' set to true");
+                return Err(Error::local_error(WrapperErrorKind::InconsistentParams));
+            }
+
+            if self.is_decryption_key
+                && rsa_scheme.algorithm() != RsaSchemeAlgorithm::RsaEs
+                && rsa_scheme.algorithm() != RsaSchemeAlgorithm::Oaep
+                && rsa_scheme.algorithm() != RsaSchemeAlgorithm::Null
+            {
+                error!("Invalid rsa scheme algorithm provided with 'restricted' set to false and 'is_decryption_key' set to true");
+                return Err(Error::local_error(WrapperErrorKind::InconsistentParams));
+            }
+        }
+
+        Ok(PublicRsaParameters {
+            symmetric_definition_object,
+            rsa_scheme,
+            key_bits,
+            exponent,
+        })
+    }
+}
+
+/// Strucure used to hold the value of a RSA exponent
+#[derive(Clone, Debug, Copy)]
+pub struct RsaExponent {
+    value: u32,
+}
+
+impl RsaExponent {
+    /// Function for creating a new RsaExponent
+    ///
+    /// # Errors
+    /// Will return an error if the value passed into the function
+    /// is not a valid RSA exponent.
+    pub fn create(value: u32) -> Result<Self> {
+        if !RsaExponent::is_valid(value) {
+            error!(
+                "Received invalid value {} when creating rsa exponent",
+                value,
+            );
+            return Err(Error::WrapperError(WrapperErrorKind::InvalidParam));
+        }
+        Ok(RsaExponent { value })
+    }
+
+    /// Method that returns the value of the rsa exponent.
+    pub const fn value(&self) -> u32 {
+        self.value
+    }
+
+    /// Private method for checking if a value is valid.
+    fn is_valid(value: u32) -> bool {
+        if 3 > value || value == 0 {
```

Comment:
*Trusted Platform Module Library 
Part 2: Structures 
Family “2.0” 
Level 00 Revision 01.59 
November 8, 2019 
Section 12.2.3.5  TPMS_RSA_PARMS *
>A TPM compatible with this specification and supporting RSA shall support two primes and an exponent 
>of zero. An exponent of zero indicates that the exponent is the default of 2^16 + 1. Support for other values 
>is optional. Use of other exponents in duplicated keys is not recommended because the resulting keys 
>would not be interoperable with other TPMs. 

*Table 196 — Definition of {RSA} TPMS_RSA_PARMS Structure *
>| exponent | U32 | The public exponent, < prime number greater than 2. 

So you are right that is wrong.

Reply:
Fixed